File: apphub/components/app_card.py
from typing import TYPE_CHECKING

from gi.repository import GLib, Gtk

if TYPE_CHECKING:
    from apphub.api.types import AppHitJson

from apphub.utils.image import load_image
from apphub.utils.navigator import navigator


@Gtk.Template(resource_path="/com/bedsteler20/AppHub/app_card.ui")
class AppCard(Gtk.Button):
    __gtype_name__ = "AppCard"

    image: Gtk.Image = Gtk.Template.Child()
    name_label: Gtk.Label = Gtk.Template.Child()
    description_label: Gtk.Label = Gtk.Template.Child()
    verified_label: Gtk.Label = Gtk.Template.Child()
    verified_box: Gtk.Box = Gtk.Template.Child()

    def do_clicked(self, *args) -> None:
        navigator.visit(self, f"/apps/{self._app_id}")

    def set_app(self, app: "AppHitJson", load_icon=True):
        self.name_label.set_label(app["name"])
        self._app_id = app["app_id"]

        if app.get("description"):
            self.description_label.set_label(app["summary"])

        if app.get("verification_verified") == "true":
            self.verified_box.set_visible(True)
            if app.get("verification_method") == "login_provider":
                if app.get("verification_login_provider") in ["gnome", "kde"]:
                    self.verified_label.set_label(
                        app["verification_login_provider"].upper()
                    )
                else:
                    self.verified_label.set_label(
                        f"@{app['verification_login_name']} on {app['verification_login_provider']}"
                    )
            elif app.get("verification_method") == "website":
                self.verified_label.set_label(app["verification_website"])
            else:
                self.verified_label.set_label("Verified")

        if load_icon:
            load_image(self.image, app["icon"])

File: src/app_card.rs
use crate::{flathub_client::AppHit, image::load_image};
use glib::{subclass::types::ObjectSubclassIsExt, Variant};
use gtk::prelude::*;

mod imp {
    use adw::subclass::prelude::*;
    use glib::subclass::InitializingObject;
    use gtk::CompositeTemplate;

    #[derive(CompositeTemplate, Default)]
    #[template(file = "src/app_card.blp")]
    pub struct ApphubAppCard {
        #[template_child]
        pub name_label: TemplateChild<gtk::Label>,
        #[template_child]
        pub description_label: TemplateChild<gtk::Label>,
        #[template_child]
        pub image: TemplateChild<gtk::Image>,
        #[template_child]
        pub verified_box: TemplateChild<gtk::Box>,
        #[template_child]
        pub verified_label: TemplateChild<gtk::Label>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for ApphubAppCard {
        const NAME: &'static str = "ApphubAppCard";
        type Type = super::ApphubAppCard;
        type ParentType = gtk::Button;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
        }

        fn instance_init(obj: &InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for ApphubAppCard {}
    impl WidgetImpl for ApphubAppCard {}
    impl ButtonImpl for ApphubAppCard {}
}

glib::wrapper! {
    pub struct ApphubAppCard(ObjectSubclass<imp::ApphubAppCard>)
    @extends gtk::Button, gtk::Widget,
    @implements gtk::Accessible, gtk::Buildable,
                gtk::ConstraintTarget, gtk::Actionable;
}

impl ApphubAppCard {
    pub fn new() -> Self {
        glib::Object::builder().build()
    }

    pub fn load_from_hit(&self, app: AppHit) {
        let ui = self.imp();
        ui.name_label.set_text(&app.name);
        ui.description_label.set_text(&app.summary);

        ui.image.set_from_icon_name(Some("image-missing"));
        self.set_action_name(Some("win.navigator.visit"));
        self.set_action_target_value(Some(&Variant::from(format!("/app/{}", &app.app_id))));
        if let Some(icon) = app.icon.as_ref() {
            load_image(icon, &ui.image);
        }
        // BUG: After navigation to a new page the button is staying focused
        // preventing the user from scrolling with the mouse wheel.
        self.set_focus_on_click(false);
    }
}
